### PR TITLE
[SOAR-6363, SOAR-6156] Major/Minor Version bump validator

### PR DESCRIPTION
**Pull request by pmara-r7**
## Description
We want to properly signal to developers when a major and minor version bumps are needed in the plugin spec (major is more important because it means breaking changes)

In general, here is what the "major" version bump is looking at:

actions:
  * any removed actions?
  * have action titles changed?
  * for each action:
    * any removed output/input?
    * have the types or titles changed for these output/input?
    * any new required inputs or existing inputs changed to required?
    * any outputs changed from required to optional?

(same above but for triggers)
types:
  * are all types still present in the spec? (top level)
  * for each type:
    * each subtype still present?
    * does each subtype have the same underlying type?

connections:
  * All connection fields still there?
  * Type/title changed for any connection field?
  * Any new, required, connection fields?

Minor Version Bump Checks:

actions/triggers:
  * New trigger/action
  * New non-required input or new output for existing trigger/action

https://issues.corp.rapid7.com/browse/SOAR-6156
https://issues.corp.rapid7.com/browse/SOAR-6363

## Testing
Unit tested (see PR for unit tests)
Ran all tests in the validator and all passed

### CODE REVIEW

**Review by mrinehart-r7 — CHANGES_REQUESTED** [review]
Few things:
1. We talk about version bumping pretty casually, but since this is community-facing we should use more proper terminology here - can you use `increment`/`decrement` instead of bumping?
2. Would it be possible to tell the user what the version string should be if it is non-compliant? Eg. instead of "this requires a major version bump, X.0.0", we tell them "this requires a major version increment to 3.0.0" (for example)?
3. W/r/t terminology, we need to make sure users know where problems are. Can you replace `spec` with `plugin.spec.yaml` in the messaging?

File: icon_validator/rules/plugin_validators/version_bump_validator.py [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+                    if _blob.name == RepoConstants.PLUGIN_SPEC:
+                        blob = _blob
+                        break
+                if blob is None:
+                    # throw error: no plugin spec found in remote
+                    raise ValidationException("No plugin spec found in remote repo")
+                break
+        if remote is None:
+            # throw exception : origin/master not found
+            raise ValidationException("Remote origin/master not found.'master' branch name changed, update validator")
```
[the remark this answers is not in the document]
Comment (pmara-r7) [reply icon_validator/rules/plugin_validators/version_bump_validator.py:60]: by "this" do you mean getting the blob object from the repo?

File: icon_validator/rules/plugin_validators/version_bump_validator.py [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+            # We already bumped the major version- skip the rest of the validation
+            return
+        else:
+            try:
+                VersionBumpValidator.validate_actions(remote_spec, local_spec)
+                VersionBumpValidator.validate_triggers(remote_spec, local_spec)
+                VersionBumpValidator.validate_connections(remote_spec, local_spec)
+                VersionBumpValidator.validate_no_types_changed(remote_spec, local_spec)
+            except ValidationException as e:
+                raise e
```
[the remark this answers is not in the document]
Comment (pmara-r7) [reply icon_validator/rules/plugin_validators/version_bump_validator.py:338]: I will change it- the reason I re-raised is I don't want to get into the minor checks unless the major checks pass. What pattern should I be using instead here?

[the remark this answers is not in the document]
Comment (pmara-r7) [reply icon_validator/rules/plugin_validators/version_bump_validator.py:336]: I guess removing it just works nvm

File: icon_validator/rules/plugin_validators/version_bump_validator.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mrinehart-r7) [reply icon_validator/rules/plugin_validators/version_bump_validator.py]: Discussed over Zoom - will move into own function

File: icon_validator/rules/plugin_validators/version_bump_validator.py [hunk 3]
```diff
@@ -32,57 +32,70 @@ class SpecConstants:
...
                 # now get the blob representing the plugin folder and loop over until we find plugin spec
-                for _blob in _remote.object.tree[RepoConstants.PLUGIN_DIRNAME][spec.spec_dictionary()["name"]]:
+                for _blob in _remote.object.tree[RepoConstants.PLUGIN_DIRNAME][plugin_name]:
                     if _blob.name == RepoConstants.PLUGIN_SPEC:
                         blob = _blob
                         break
                 if blob is None:
                     # throw error: no plugin spec found in remote
-                    raise ValidationException("No plugin spec found in remote repo")
+                    raise ValidationException(f"{RepoConstants.PLUGIN_SPEC} found in remote repo")
```
Comment (mrinehart-r7) [inline icon_validator/rules/plugin_validators/version_bump_validator.py:75]:
```suggestion
                    raise ValidationException(f"{RepoConstants.PLUGIN_SPEC} not found in remote repo")
```

File: icon_validator/rules/plugin_validators/version_bump_validator.py [hunk 4]
```diff
@@ -94,13 +107,16 @@ def validate_no_types_changed(remote, local):
                     local_type_in = local[SpecConstants.TYPES][type_key]
                     if type_inner_key not in local_type_in:
                         raise ValidationException(f"Type {type_inner_key} removed from {type_key} without a major"
-                                                  f" version bump. You may add it back or bump the major version")
+                                                  f" version increment."
+                                                  f"{VersionBumpValidator.MAJOR_INSTRUCTIONS_STRING}")
                     if type_inner_val[SpecConstants.TYPE] != local_type_in[type_inner_key][SpecConstants.TYPE]:
                         raise ValidationException(f"Type {type_inner_key} changed in type {type_key} without a major"
-                                                  f"version bump. You may add it back or bump the major version")
+                                                  f"version increment."
```
Comment (mrinehart-r7) [inline icon_validator/rules/plugin_validators/version_bump_validator.py:114]:
```suggestion
                                                  f" version increment."
```

**Review by mrinehart-r7 — CHANGES_REQUESTED** [review]

**Review by mrinehart-r7 — APPROVED** [review]

File: icon_validator/rules/plugin_validators/version_bump_validator.py [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+            # ... to accept checking functions for minor/major bump
+        if not VersionBumpValidator.check_minor_version_bump_needed(remote_spec, local_spec):
+            # already validated that no major bump was needed and we bumped minor version- skip further validation
+            return
+        else:
+            try:
+                VersionBumpValidator.validate_minor_triggers(remote_spec, local_spec)
+                VersionBumpValidator.validate_minor_actions(remote_spec, local_spec)
+            except ValidationException as e:
+                raise e
```
[the remark this answers is not in the document]
Comment (pmara-r7) [reply icon_validator/rules/plugin_validators/version_bump_validator.py]: Removed try catch

**Review by mberezin-r7 — COMMENTED** [review]
We should also print which action/trigger/connection a particular violation corresponds to.

File: icon_validator/rules/plugin_validators/version_bump_validator.py [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+
+    def validate_no_inner_type_changes(self, remote, local):
+        self.abstract_validate_no_change(remote, local, SpecConstants.INPUT, SpecConstants.TYPE)
+        self.abstract_validate_no_change(remote, local, SpecConstants.OUTPUT, SpecConstants.TYPE)
+
+    def validate_no_titles_changed(self, remote, local):
+        self.abstract_validate_no_change(remote, local, SpecConstants.INPUT, SpecConstants.TITLE)
+        self.abstract_validate_no_change(remote, local, SpecConstants.OUTPUT, SpecConstants.TITLE)
+
+    def abstract_validate_no_change(self, remote, local, input_or_output, field):
```
Comment (mberezin-r7) [inline icon_validator/rules/plugin_validators/version_bump_validator.py:129]: Type hinting here would help for readability. This is a string but corresponds to only one of two possible values.

File: icon_validator/rules/plugin_validators/version_bump_validator.py [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+                                          f"a major version increment.{self.MAJOR_INSTRUCTIONS_STRING}")
+
+    def validate_no_action_removed(self, remote, local):
+        for action in remote[SpecConstants.ACTIONS]:
+            if action not in local[SpecConstants.ACTIONS]:
+                raise ValidationException("Action has been removed from spec without a major version increment."
+                                          f"{self.MAJOR_INSTRUCTIONS_STRING}")
+
+    def validate_no_trigger_removed(self, remote, local):
+        for action in remote[SpecConstants.TRIGGERS]:
```
Comment (mberezin-r7) [inline icon_validator/rules/plugin_validators/version_bump_validator.py:184]: This should be named `triggers` variable.

File: icon_validator/rules/plugin_validators/version_bump_validator.py [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+            self.validate_no_input_new_or_required(remote, local)
+        if SpecConstants.OUTPUT in remote and SpecConstants.OUTPUT in local:
+            self.validate_no_output_removed(remote, local)
+            self.validate_no_output_no_longer_required(remote, local)
+
+        self.validate_no_titles_changed(remote, local)
+        self.validate_no_inner_type_changes(remote, local)
+
+    def check_major_version_increment_needed(self, remote, local):
+        local_version = local["version"].split('.')
```
Comment (mberezin-r7) [inline icon_validator/rules/plugin_validators/version_bump_validator.py:252]: Check for hyphen semver strings with patch suffixes.

**Review by mberezin-r7 — APPROVED** [review]
Approved pending couple comments.

File: icon_validator/rules/plugin_validators/version_bump_validator.py (hunk 7 above)
Comment (pmara-r7) [reply icon_validator/rules/plugin_validators/version_bump_validator.py]: done

File: icon_validator/rules/plugin_validators/version_bump_validator.py (hunk 8 above)
Comment (pmara-r7) [reply icon_validator/rules/plugin_validators/version_bump_validator.py:260]: Made a function that takes everything before a hyphen (if a hyphen is in the last slot)

File: icon_validator/rules/plugin_validators/version_bump_validator.py (hunk 6 above)
Comment (pmara-r7) [reply icon_validator/rules/plugin_validators/version_bump_validator.py]: type hinting added pretty much everywhere now

File: icon_validator/rules/plugin_validators/version_bump_validator.py (hunk 8 above)
Comment (pmara-r7) [reply icon_validator/rules/plugin_validators/version_bump_validator.py:260]: Added

File: icon_validator/rules/plugin_validators/version_bump_validator.py (hunk 7 above)
Comment (pmara-r7) [reply icon_validator/rules/plugin_validators/version_bump_validator.py]: done

**Review by mberezin-r7 — DISMISSED** [review]

File: icon_validator/rules/plugin_validators/version_bump_validator.py [hunk 9]
```diff
@@ -266,10 +276,14 @@ def check_major_version_increment_needed(self, remote, local):
...
 
-    def check_minor_version_increment_needed(self, remote, local):
+    def check_minor_version_increment_needed(self, remote: dict, local: dict):
+        # input: complete spec dictionary
+        # Checks to see if version is valid sem-ver, and if we already bumped minor version
         local_version = local["version"].split('.')
         remote_version = remote["version"].split('.')
         if len(local_version) == 3 and len(remote_version) == 3:
+            VersionBumpValidator.modify_version_array(local_version)
+            VersionBumpValidator.modify_version_array(remote_version)
```
Comment (mrinehart-r7) [inline icon_validator/rules/plugin_validators/version_bump_validator.py:286]: Did you mean to assign these function calls to variables? As it stands right now, these will not mutate `local_version` or `remote_version` so the semver with metadata will still pose an issue.

Comment (pmara-r7) [reply icon_validator/rules/plugin_validators/version_bump_validator.py]: I did, fixed in next commit

**Review by mrinehart-r7 — APPROVED** [review]

**Review by mberezin-r7 — APPROVED** [review]

**Pull request merged**